src/Cardano/AddressV3.cpp

```diff
@@ -0,0 +1,234 @@
+// Copyright © 2017-2020 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "AddressV3.h"
+#include "AddressV2.h"
+#include "../Data.h"
+#include "../Bech32.h"
+#include "../Base32.h"
+#include "../Crc.h"
+#include "../HexCoding.h"
+#include "../Hash.h"
+
+#include <array>
+
+using namespace TW;
+using namespace TW::Cardano;
+using namespace std;
+
+bool AddressV3::parseAndCheckV3(const std::string& addr, Discrimination& discrimination, Kind& kind, Data& key1, Data& key2) {
+    auto bech = Bech32::decode(addr);
+    if (bech.second.size() == 0) {
+        // empty Bech data
+        return false;
+    }
+    // Bech bits conversion
+    Data conv;
+    auto success = Bech32::convertBits<5, 8, false>(conv, bech.second);
+    if (!success) {
+        return false;
+    }
+    if (conv.size() != 33 && conv.size() != 65) {
+        return false;
+    }
+    discrimination = (Discrimination)((conv[0] & 0b10000000) >> 7);
+    kind = (Kind)(conv[0] & 0b01111111);
+    if (kind <= Kind_Sentinel_Low || kind >= Kind_Sentinel_High) {
+        return false;
+    }
+    if ((kind == Kind_Group && conv.size() != 65) ||
+        (kind != Kind_Group && conv.size() != 33)) {
+        return false;
+    }
+
+    switch (kind) {
+        case Kind_Single:
+        case Kind_Account:
+        case Kind_Multisig:
+            assert(conv.size() == 33);
+            key1 = Data(32);
+            std::copy(conv.begin() + 1, conv.begin() + 33, key1.begin());
+            return true;
+
+        case Kind_Group:
+            assert(conv.size() == 65);
+            key1 = Data(32);
+            key2 = Data(32);
+            std::copy(conv.begin() + 1, conv.begin() + 33, key1.begin());
+            std::copy(conv.begin() + 33, conv.begin() + 65, key2.begin());
+            return true;
+
+        default:
+            return false;
+    }
+}
+
+bool AddressV3::isValid(const std::string& addr) {
+    try {
+        Discrimination discrimination;
+        Kind kind;
+        Data key1;
+        Data key2;
+        if (parseAndCheckV3(addr, discrimination, kind, key1, key2)) {
```

Comment:
it won't throw exception?

Reply:
Bech method, invoked by parseAndCheckV3, may throw if format is not correct.  I have moved the catch one method higher.